controllers/companiesController.js

```diff
@@ -0,0 +1,186 @@
+'use strict';
+
+var atob = require('atob');
+var bcrypt = require('bcrypt');
+const jwt = require('jsonwebtoken');
+
+const Company = require('../models/company');
+
+exports.signUp = async ctx => {
+  const userData = ctx.request.body;
+  let user = await Company.findOne({username: userData.username});
+
+  const incompleteBody = !userData.company_name || !userData.username || !userData.email || !userData.password;
+
+  if (user) {
+    ctx.status = 400;
+    const error_message = 'Username already exist!';
+    console.log(error_message);
+    ctx.body = {
+      errors: [
+        error_message
+      ]
+    };
+  } else if (incompleteBody) {
+    ctx.status = 400;
+    const error_message = 'Incomplete body';
+    console.log(error_message);
+    ctx.body = {
+      errors: [
+        error_message
+      ]
+    };
+  } else {
+    const saltRounds = 10;
+    const plaintextPsw = userData.password;
+    const hashPsw = await bcrypt.hash(plaintextPsw, saltRounds);
+
+    const company = {
+      company_name: userData.company_name,
+    	email: userData.email,
+    	username: userData.username,
+    	password: hashPsw,
+    	fleet:[]
+    }
+
+    try {
+      const response = await Company.create(company);
+      ctx.body = {
+        company_name: response.company_name,
+        username: response.username,
+        email: response.email
+      };
+      ctx.status = 201;
+    } catch (e) {
+      console.error(e);
+      ctx.status = 500;
+      ctx.body = {
+        message: e.message
+      };
+    }
+  }
+};
+
+exports.signIn = async ctx => {
+  if (!ctx.headers['authorization']) {
+    ctx.status = 400;
+    const error_message = 'Basic authorization in header is missing';
+    console.log(error_message);
+    ctx.body = {
+			errors: [
+				error_message
+			]
+		};
+    return;
+  }
+  const b64 = atob(ctx.headers['authorization'].split(' ').pop());
+  const [username, passwordReceived] = b64.split(':');
+  const company = await Company.findOne({username: username});
+  if (company) {
+    const areCompatible = await bcrypt.compare(passwordReceived, company.password);
+    if (areCompatible) {
+      const payload = {
+        username: company.username,
+        password: company.password
+      }
+      ctx.status = 200;
+      ctx.body = {
+        username: company.username,
+        json_token: jwt.sign(payload, "$secretword"),
+        company_name: company.company_name,
+        email: company.email
+      }
+    } else {
+      ctx.status = 401;
+      const error_message = 'Unauthorized';
+      console.log(error_message);
+      ctx.body = {
+  			errors: [
+  				error_message
+  			]
+  		};
+    }
+  } else {
+    ctx.status = 404;
+    const error_message = 'Username not found';
+    console.log(error_message);
+    ctx.body = {
```

Comment:
Again the error. I challenge you to implement a new type of error, a clas `HTTPError` that extends from `Error` and accepts a status (can be defaulted) and a message.